### PR TITLE
Add CAP tool-hanging environments for the bimanual YAM

### PR DESCRIPTION
**Pull request by viiik-inside**
## Summary
Add the eight CAP tool-hanging scenes to Arena

## Detailed description
- Port Isaac-cap's wrench, scissors, screwdriver, and pliers scenes (easy and medium) as graph YAML files sharing one Newton profile through `external_yaml`; scenes run directly through the experiment runner.
- Register the tools, pegboard fixtures, and YAM table as `LibraryObject`s, a tool-hanging YAM embodiment that subclasses the shared bimanual YAM (home pose, gripper gains, AUTOLab camera rig), and a `ToolHangingTask` that composes loop-on-rod or box-containment geometry with fixture contact and settling; it holds no robot or policy state.
- The easy wrench goal uses the same enclosing test as the medium scene; the original `threaded` radius cannot be met by a resting wrench with these collision hulls.
- Validated: all eight scenes build, place, and stay stable for three 600-step zero-action episodes with and without cameras; tools teleported onto their fixtures trigger success in every family; `test_tool_hanging_task.py` covers the goal geometry and the wrench hang.

wrench_medium

### CODE REVIEW

File: isaaclab_arena_environments/isaac_cap/tool_hanging/geometry.py [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+        assert all(lo < hi for lo, hi in zip(self.minimum_xyz, self.maximum_xyz, strict=True)), "Box must be ordered."
+
+    @classmethod
+    def from_dict(cls, goal: dict) -> PointInBox:
+        """Read the ``containment`` mapping."""
+        return cls(**goal["containment"])
+
+    def evaluate(self, T_W_T: torch.Tensor, T_W_X: torch.Tensor) -> torch.Tensor:
+        """Return one Boolean per environment from the tool pose ``T_W_T`` and fixture pose ``T_W_X``."""
+        offset = _points_w(_as_tensor((self.point_xyz,), T_W_T), T_W_T)[:, 0] - T_W_X[:, :3]
```
[the remark this answers is not in the document]
Comment (viiik-inside) [reply isaaclab_arena_environments/isaac_cap/tool_hanging/geometry.py:101]: Deliberate: this mirrors how the upstream benchmark scores its containment regions (world-axis bounds translated to the fixture origin), so the YAML numbers stay comparable with the source task. These fixtures are fixed anchors. I made that explicit in the PointInBox docstring in a3f6c24.

File: isaaclab_arena_environments/isaac_cap/tool_hanging/geometry.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    loops: tuple[Loop, ...]
+    rod: Rod
+
+    @classmethod
+    def from_dict(cls, goal: dict) -> LoopOnRod:
+        """Read the ``loop``, ``rod``, and optional ``alternative_loops`` keys."""
+        loops = (Loop(**goal["loop"]), *(Loop(**loop) for loop in goal.get("alternative_loops", [])))
+        return cls(loops=loops, rod=Rod(**goal["rod"]))
+
+    def evaluate(self, T_W_T: torch.Tensor, T_W_X: torch.Tensor) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (viiik-inside) [reply isaaclab_arena_environments/isaac_cap/tool_hanging/geometry.py:65]: The enclosing test is the upstream benchmark's predicate for these goals, and the 18.2 mm radius is its medium-scene value; success also requires fixture contact and the tool at rest for 60 steps. The hooks sit about 40 cm above the table, so a tool can only be in sustained contact with one while hanging from it. The wrench-easy scene originally used a threaded test with a 10.2 mm radius; a resting wrench's ring centre settles 9.3 mm below the shank with these collision hulls, so it never passed. Happy to add a normal term if we want tighter geometry later.

File: isaaclab_arena_environments/isaac_cap/tool_hanging/task.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+
+class ToolHangingTask(TaskBase):
+    """Hang or stow every listed tool on its fixture and hold that state for consecutive steps."""
+
+    def __init__(
+        self,
+        tools: list[Asset],
+        fixtures: list[Asset],
+        goals: list[dict],
```
[the remark this answers is not in the document]
Comment (viiik-inside) [reply isaaclab_arena_environments/isaac_cap/tool_hanging/task.py:69]: Done in a3f6c24: goals now pair with tools and fixtures by position, so each asset is named once and the sync assert is gone.

File: isaaclab_arena_environments/isaac_cap/tool_hanging/assets.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+class Fixture(LibraryObject):
+    """Kinematic pegboard fixture; its USD authors ``physics:kinematicEnabled``."""
+
+    tags = ["object", "fixture"]
+
+    def __init__(self, initial_pose: Pose | None = None, collision_mode: str = "mesh", **kwargs):
+        super().__init__(initial_pose=initial_pose, collision_mode=collision_mode, **kwargs)
+
+    def get_world_bounding_box(self) -> AxisAlignedBoundingBox:
```
[the remark this answers is not in the document]
Comment (viiik-inside) [reply isaaclab_arena_environments/isaac_cap/tool_hanging/assets.py]: Agreed. a3f6c24 makes PlacementAsset.get_world_bounding_box use enclosing_after_rotation (exact for quarter turns, conservative otherwise) and removes the override; the existing bounding-box and placer tests still pass.

File: isaaclab_arena_environments/isaac_cap/tool_hanging/wrench_easy.yaml [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  registry_name: tool_hanging_bimanual_yam
+  params:
+    left_mount_position: [0.3025, 0.17, 0.81]
+    right_mount_position: [0.3025, -0.17, 0.81]
+background:
+  id: background
+  registry_name: yam_workcell_table
+  params:
+    initial_pose:
+      position_xyz: [0.2525, 0.0, 0.75]
```
[the remark this answers is not in the document]
Comment (viiik-inside) [reply isaaclab_arena_environments/isaac_cap/tool_hanging/wrench_easy.yaml:19]: These are the layouts as validated in Isaac-cap: three easy scenes keep the upstream Level-1 table pose (0.728) and the easy wrench keeps its originally verified rack height; the medium scenes share the Level-2 layout. Kept as-is so the scored scenes stay identical to the source rather than unifying them here.

File: isaaclab_arena_environments/isaac_cap/tool_hanging/screwdriver_medium.yaml [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+- id: pegboard_rack
+  registry_name: pegboard_rack
+  params:
+    initial_pose:
+      position_xyz: [0.88, -0.16, 1.12800769]
+      rotation_xyzw: [0.0, 0.0, 0.0, 1.0]
+- id: empty_warehouse_hdr_light
+  registry_name: empty_warehouse_dome_light
+- id: rack
+  registry_name: screwdriver_box_upstream
```
[the remark this answers is not in the document]
Comment (viiik-inside) [reply isaaclab_arena_environments/isaac_cap/tool_hanging/screwdriver_medium.yaml:34]: The success volume comes from the upstream task definition and is the same for both scenes; the two meshes differ only in wall geometry (the medium one was measured from AUTOLab's stored mesh in Isaac-cap #229). Unifying the easy scene onto that mesh would change a validated scene, so both stay for now.

File: isaaclab_arena_environments/isaac_cap/tool_hanging/task.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+
+def goal_geometry_satisfied(env, tool_name: str, fixture_name: str, geometry: LoopOnRod | PointInBox) -> torch.Tensor:
+    """Evaluate the goal geometry on the live tool and fixture poses."""
+    return geometry.evaluate(env.arena_world.get_pose_w(tool_name), env.arena_world.get_pose_w(fixture_name))
+
+
+def contact_force_above(env, sensor_name: str, force_threshold: float) -> torch.Tensor:
+    """Check the strongest filtered contact force reported by a sensor."""
+    force_w = env.scene[sensor_name].data.force_matrix_w.torch
```
[the remark this answers is not in the document]
Comment (viiik-inside) [reply isaaclab_arena_environments/isaac_cap/tool_hanging/task.py]: Done in a3f6c24.

File: isaaclab_arena/tests/test_tool_hanging_task.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            "point_xyz": [0.0, 0.0, 0.05],
+        }
+    })
+    assert goal.evaluate(T_W_T, T_W_X).tolist() == [True]
+    goal.point_xyz = (0.0, 0.0, 0.2)
+    assert goal.evaluate(T_W_T, T_W_X).tolist() == [False]
+
+
+def _test_wrench_hangs_on_hook(_simulation_app):
+    import torch
```
[the remark this answers is not in the document]
Comment (viiik-inside) [reply isaaclab_arena/tests/test_tool_hanging_task.py]: Done in a3f6c24.

File: isaaclab_arena_environments/isaac_cap/tool_hanging/cameras.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+            focus_distance=28.0,
+            horizontal_aperture=6.4,
+            vertical_aperture=_VERTICAL_APERTURE,
+            clipping_range=clipping_range,
+        ),
+        offset=CameraCfg.OffsetCfg(pos=position, rot=rotation_xyzw, convention="ros"),
+    )
+
+
+def _fixed_camera(name: str, position, rotation_xyzw) -> CameraCfg:
```
[the remark this answers is not in the document]
Comment (viiik-inside) [reply isaaclab_arena_environments/isaac_cap/tool_hanging/cameras.py]: Done in a3f6c24.

File: isaaclab_arena_environments/isaac_cap/tests/test_tool_hanging.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
+# Copyright (c) 2026, The Isaac Lab Arena Project Developers (https://github.com/isaac-sim/IsaacLab-Arena/blob/main/CONTRIBUTORS.md).
```
Comment (qianl-nv) [inline isaaclab_arena_environments/isaac_cap/tests/test_tool_hanging.py:1]: please move into isaaclab_arena_environments/isaac_cap/tests/test_temporary_envs.py
Peter's intention was to keep all cap-related unit test in the isaaclab_arena_environments/isaac_cap/tests folder so we can delete cleaning
https://github.com/isaac-sim/IsaacLab-Arena/pull/1263 
it leaks through in the syringe PR (my bad), I'll fix that.

File: isaaclab_arena_environments/isaac_cap/tool_hanging/cameras.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+@configclass
+class ToolHangingYamCameraCfg(BimanualYamCameraCfg):
+    """Fixed overhead and side views plus both wrist cameras, named as CAP's policy expects."""
+
+    top_camera: CameraCfg = _fixed_camera("top_camera", _TOP_POSITION, _TOP_ROTATION_XYZW)
+    side_camera: CameraCfg = _fixed_camera("side_camera", _SIDE_POSITION, _SIDE_ROTATION_XYZW)
+    left_wrist_camera: CameraCfg = _wrist_camera("LeftRobot", "left_wrist_camera")
+    right_wrist_camera: CameraCfg = _wrist_camera("RightRobot", "right_wrist_camera")
+
+    def set_robot_mount_positions(self, left: tuple[float, float, float], right: tuple[float, float, float]) -> None:
```
Comment (qianl-nv) [inline isaaclab_arena_environments/isaac_cap/tool_hanging/cameras.py:84]: this is no-op, can delete

**Review by qianl-nv — APPROVED** [review]
LGTM thanks! Small comments, mainly for moving unit test. Feel free to merge after fixing.

File: isaaclab_arena/relations/placement_asset.py [hunk 12]
```diff
@@ -153,16 +152,16 @@ def get_bounding_box(self) -> AxisAlignedBoundingBox:
...
 
-        Unset, ranged, and per-environment poses leave the root-relative bounds unchanged.
+        Exact for quarter-turn Z rotations and conservative for any other rotation. Unset, ranged,
+        and per-environment poses leave the root-relative bounds unchanged.
         """
         bounding_box = self.get_bounding_box()
         initial_pose = self.get_initial_pose()
         if not isinstance(initial_pose, Pose):
             return bounding_box
-        quarters = quaternion_to_90_deg_z_quarters(initial_pose.rotation_xyzw)
```
Comment (qianl-nv) [inline isaaclab_arena/relations/placement_asset.py:164]: @alexmillane @cvolkcvolk not sure if placement solver intentionally limit this rotation to 90 degree for AABB alignment enforcement. Do you think this change allowing for all angles is ok?

File: isaaclab_arena_environments/isaac_cap/tests/test_tool_hanging.py (hunk 10 above)
Comment (viiik-inside) [reply isaaclab_arena_environments/isaac_cap/tests/test_tool_hanging.py:1]: Moved in 0f00c2e to isaaclab_arena_environments/isaac_cap/tests/test_tool_hanging.py with the isaac_cap marker. I kept it as its own file rather than appending to test_temporary_envs.py because that module marks everything with_subprocess, while this test uses the in-process persistent SimulationApp. It passes under the CI job's `-m isaac_cap` filter.

File: isaaclab_arena_environments/isaac_cap/tool_hanging/cameras.py (hunk 11 above)
Comment (viiik-inside) [reply isaaclab_arena_environments/isaac_cap/tool_hanging/cameras.py:84]: It is doing something: the shared rig's __init__ calls set_robot_mount_positions and moves the top camera to the arm-mount midpoint plus an offset (about 0.64, 0, 1.75 here), whereas AUTOLab's fixed cameras are authored in world coordinates (0.30, 0, 1.67). The override keeps the parent from relocating them. I reworded the docstring in 0f00c2e so that is clear.

File: isaaclab_arena/relations/placement_asset.py (hunk 12 above)
Comment (viiik-inside) [reply isaaclab_arena/relations/placement_asset.py:164]: For context: the previous code asserted (hard failure) on any rotation that was not a quarter turn about Z, and the pitched hooks in the medium scenes hit that assert. enclosing_after_rotation gives the identical box for quarter turns and a conservative (larger) box otherwise, so no existing placement changes and the placer only becomes more permissive about which anchors it accepts. The existing bounding-box and placer tests pass. If you would rather keep the core assert, I can drop this and restore the package-local override instead; either is a two-line change.

Comment (zhx06) [reply isaaclab_arena/relations/placement_asset.py:164]: Thanks for your efforts! I think we should keep it package-local override, since other functions in the main repo are heavily relying on this bounding box behavior.

Comment (viiik-inside) [reply isaaclab_arena/relations/placement_asset.py:164]: Reverted the core change in aa0401f09; the pitched-hook handling is back as a local override on the tool-hanging Fixture class, so this PR no longer touches isaaclab_arena/relations. Tests and the medium-scene builds pass.

**Comment by viiik-inside** [comment]
Rebased onto main as one commit. Adapted to the new APIs: the task now returns a TaskTerminationCfg with a single ProgressObjective, and the embodiment passes its rig through the shared YAM's new camera_config parameter, so the hook in the cable-routing embodiment and the camera override are gone. The pitched-hook bounds stay as a local override on the tool-hanging Fixture; the PR no longer touches isaaclab_arena/. Tests, all eight scene builds, and the camera rig re-verified in the container.

**Comment by viiik-inside** [comment]
The remaining red job ("Run tests with subprocess") is `test_rsl_rl.py::test_rl_train_and_eval_lift_object`, and it fails the same way on `main` itself: runs 35547174221 and 35567778961 (4dc0cf5) and 35578966418 (9886659) all fail on exactly that test. The child `train.py` either crashes in Kit at startup or times out at 900 s before any Arena code runs, and this PR does not touch the training path (it only adds a registration call and the graph-YAML registration in `cli.py`). All other jobs on this head pass, including the Isaac CAP sanity checks.
